---
title: "Feature Barcode QC"
author: "Robin Meyers"
date: "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
params:
    rmd: "report/featurebarcode-qc-report.Rmd"
   
---

## Source
<a download="featurebarcode-qc-report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>

```{r include=FALSE}
# threads: "`r snakemake@threads`"
# library_fasta: "`rsnakemake@config$library_fasta`"
# read_stats: "`r snakemake@input$stats`"
# feature_counts: "`r snakemake@input$counts`"
# target_regex: "`r snakemake@config$target_regex`"
knitr::opts_chunk$set(echo = T, message=F, warning=F)

format_si <- function(...) {
    function(x) {
        limits <- c(1e-24, 1e-21, 1e-18, 1e-15, 1e-12,
                    1e-9,  1e-6,  1e-3,  1e0,   1e3,
                    1e6,   1e9,   1e12,  1e15,  1e18,
                    1e21,  1e24)
        prefix <- c("y",   "z",   "a",   "f",   "p",
                    "n",   "u",   "m",   "",   "k",
                    "M",   "G",   "T",   "P",   "E",
                    "Z",   "Y")
        prefix <- ifelse(prefix != "", paste0(" ", prefix), "")
        
        # Vector with array indices according to position in intervals
        i <- findInterval(abs(x), limits, rightmost.closed = T)
        
        # Set prefix to " " for very small values < 1e-24
        i <- ifelse(i == 0, which(limits == 1e0), i)
        
        paste0(format(round(x/limits[i], 1),
                      trim = TRUE, scientific = FALSE, ...),
               prefix[i])
    }
}

pretty_log10_breaks <- function (scalers = NULL, log_breaks = NULL) {
    function(x) {
        top_10 <- log10(x[2])
        bottom_10 <- log10(x[1])
        
        n_logs <- top_10 - bottom_10
        # cat("len(x)", length(x), "\n")
        # cat("top:", 10^top_10, "\n")
        # cat("bottom:", 10^bottom_10, "\n")
        # cat("n logs:", top_10 - bottom_10, "\n")
        if (is.null(log_breaks)) {
            log_breaks <- ceiling(n_logs / 6)
        }
        # cat("log breaks:", log_breaks, "\n")
        if (is.null(scalers)) {
            if (log_breaks > 1) {
                scalers <- 1
            } else if (n_logs == 1) {
                scalers <- 1:9
            } else if (n_logs < 2.5) {
                scalers <- c(1, 2, 5)    
            } else if (n_logs < 3.5) {
                scalers <- c(1, 3)
            } else {
                scalers <- 1
            }
        }
        # cat("scalers:", scalers, "\n")
        
        breaks = 10^rep(seq(floor(bottom_10), ceiling(top_10), by=log_breaks),
                        each = length(scalers)) * scalers
    }
}

pretty_log10_limits <- function(x) {
    max_x <- log10(max(x))
    options_upper_x <- c(10^floor(max_x), 2*10^(floor(max_x)),
                         5*10^floor(max_x), 10^ceiling(max_x))
    upper_x <- min(options_upper_x[options_upper_x >= 10^max_x])
    min_x <- log10(min(x))
    options_lower_x <- c(10^floor(min_x), 2*10^(floor(min_x)),
                         5*10^floor(min_x), 10^ceiling(min_x))
    lower_x <- max(options_lower_x[options_lower_x <= 10^min_x])
    return(c(lower_x, upper_x))
}

scale_y_log10_pretty <- function (...)
{
    
    scale_args <- list(trans = scales::log10_trans())
    
    if (!hasArg(labels)) {
        scale_args$labels <- format_si()
    }
    if (!hasArg(breaks)) {
        scale_args$breaks <- pretty_log10_breaks()
    }
    
    do.call(scale_y_continuous, c(scale_args, list(...)))       
    
}

scale_x_log10_pretty <- function (...)
{
    
    scale_args <- list(trans = scales::log10_trans())
    
    if (!hasArg(labels)) {
        scale_args$labels <- format_si()
    }
    if (!hasArg(breaks)) {
        scale_args$breaks <- pretty_log10_breaks()
    }
    
    do.call(scale_x_continuous, c(scale_args, list(...)))
}

```


### Setup

```{r setup}
### Load other libraries
library(plyr)
library(Biostrings)
library(magrittr)
library(tidyverse)
library(scales)
library(cowplot)
library(ggrepel)


threads <- snakemake@threads

if (threads > 1) {
    library(doMC)
    registerDoMC(cores=threads)
    do_parallel <- T
} else {
    do_parallel <- F
}

```

### Load data

```{r 'load data', fig.height=5, fig.width=8}

samplesheet <- read_csv(snakemake@config$samplesheet)
read_stats_raw <- read_csv(snakemake@input$stats)
feature_counts <- read_tsv(snakemake@input$counts)

```


```{r}

read_stats <- read_stats_raw %>% gather(step, count, -name, -id) %>% 
    group_by(name) %>% 
    mutate(percent_raw = count / count[step=="reads"])

step_order <- colnames(read_stats_raw %>% select(-name, -id))

```


## Read Stats

### Total Reads

```{r fig.width=8, fig.height=4}

message("read stats")

pdna_read_stats <- read_stats %>% 
    filter(name == "pDNA")

sample_read_stats <- read_stats %>% 
    filter(name != "pDNA")

pdna_gg <- pdna_read_stats %>% 
    ggplot(aes(x=factor(step, levels=step_order),
               y=count, group=name)) +
    geom_line(show.legend=F) +
    geom_point(show.legend=F) +
    annotation_logticks(sides = 'l') +
    scale_y_log10_pretty(
        limits  = pretty_log10_limits(pdna_read_stats$count)) +
    labs(y = "Count",
         title = "pDNA reads") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle=45, vjust=1, hjust=1),
          legend.title = element_blank())


sample_gg <- sample_read_stats %>% 
    ggplot(aes(x=factor(step, levels=step_order),
               y=count, group=name, color=name)) +
    geom_line(show.legend=F) +
    geom_point(show.legend=F) +
    annotation_logticks(sides = 'l') +
    scale_y_log10_pretty(
        limits  = pretty_log10_limits(sample_read_stats$count)) +
    labs(y = "Count",
         title = "Sample reads") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle=45, vjust=1, hjust=1),
          legend.title = element_blank())

plot_grid(pdna_gg, sample_gg)

```

### Percent of raw reads

```{r fig.width=8, fig.height=4}

pdna_gg <- 
    read_stats %>% 
    filter(name == "pDNA") %>%
    ggplot(aes(x=factor(step, levels = step_order),
               y=percent_raw, group=name)) +
    geom_line() +
    geom_point() +
    scale_y_continuous(labels = percent_format(accuracy=1),
                       breaks=pretty_breaks(5),
                       limits = range(read_stats$percent_raw)) +
    labs(y = "Percent of raw reads",
         title = "pDNA reads") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle=45, vjust=1, hjust=1))

sample_gg <- 
    read_stats %>% 
    filter(name != "pDNA") %>%
    ggplot(aes(x=factor(step, levels = step_order),
               y=percent_raw, group=name, color=name)) +
    geom_line(show.legend=F) +
    geom_point(show.legend=F) +
    scale_y_continuous(labels = percent_format(accuracy=1),
                       breaks=pretty_breaks(5),
                       limits = range(read_stats$percent_raw)) +
    labs(y = "Percent of raw reads",
         title = "Sample reads") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle=45, vjust=1, hjust=1))


plot_grid(pdna_gg, sample_gg)


```



## Number of features per gene target in pDNA

```{r fig.width=4, fig.height=4}

sgrna_lib <- readDNAStringSet(snakemake@config$library_fasta) %>%
    names() %>%
    tibble(feature = .) %>%
    mutate(target = str_match(feature, snakemake@config$target_regex)[,2])


pDNA_counts <- feature_counts %>%
    filter(sample == "pDNA") %>%
    left_join(sgrna_lib, .) %>%
    mutate(count = ifelse(is.na(count), 0, count))

num_features_per_target <- pDNA_counts %>%
    filter(!is.na(target)) %>%
    group_by(target) %>%
    summarise(features = sum(count > 0))

num_features_per_target %>%
    ggplot() +
    geom_histogram(aes(x = features, y = ..count..), binwidth = 1) +
    geom_text(stat = "bin", binwidth = 1, vjust = -0.25,
              aes(x = features, y = ..count.., label = ifelse(..count.. > 0, ..count.., ""))) +
    scale_x_continuous(breaks = 0:max(num_features_per_target$features),
                       limits = c(-0.5, max(num_features_per_target$features)+0.5)) +
    labs(x = "Number of features per target",
         y = "Targets")

```

## Distribution of pDNA features

```{r fig.width=6, fig.height=5}
message("pdna distribution")

quantiles <- c(0.1, 0.25, 0.5, 0.75, 0.9)
pDNA_quantiles <- quantile(pDNA_counts$count, quantiles)

h_segments <- tibble(x = min(pDNA_counts$count[pDNA_counts$count > 0]), xend = pDNA_quantiles,
                     y = quantiles, yend=quantiles)
v_segments <- tibble(y = 0, yend = quantiles,
                     x = pDNA_quantiles, xend = pDNA_quantiles)
pDNA_quantile_segments <- bind_rows(h_segments, v_segments)
pDNA_quantile_labels <- tibble(x = pDNA_quantiles, y = quantiles,
                               label = paste0(names(pDNA_quantiles), " = ", round(pDNA_quantiles)))

pDNA_counts %>% 
    mutate(ecdf = ecdf(count)(count)) %>% 
    filter(count > 0) %>% 
    ggplot(aes(x=count, y=ecdf)) +
    geom_segment(aes(x=x,xend=xend,y=y,yend=yend),
                 data=pDNA_quantile_segments, lty=2, color="grey70") +
    geom_text(aes(x=x, y=y, label=label), vjust=-0.5, hjust=1,
              data = pDNA_quantile_labels) +
    geom_line() +
    scale_y_continuous(limits=c(0,1), breaks=pretty_breaks(5)) +
    annotation_logticks(sides='b') +
    scale_x_log10_pretty(limits = c(min(pDNA_counts$count[pDNA_counts$count > 0]),
                                    max(pDNA_counts$count))) +
    labs(x="pDNA feature counts", y="Cumulative fraction")


```

## Feature barcode UMIs per cell

### UMI counts by cell rank

```{r fig.width = 9, fig.height = 3}
message("umis per cell")

samples_per_row <- 4

cell_count_summary <- 
    feature_counts %>% 
    filter(sample != "pDNA") %>%
    group_by(sample, cell) %>% 
    summarise(count = sum(count)) %>% 
    group_by(sample) %>% 
    mutate(rank = row_number(-count)) %>%
    ungroup() %>%
    mutate(plot_group = ceiling(as.numeric(factor(sample))/samples_per_row))

gg_list <- dlply(cell_count_summary, .(plot_group),
                 function(df) {
                     gg <- df %>% 
                         group_by(sample, count) %>% 
                         filter(rank == min(rank) | rank == max(rank)) %>% 
                         ggplot(aes(x = rank, y = count)) +
                         facet_wrap(~ sample, nrow = 1, scales = "free") +
                         geom_line() +
                         scale_x_log10_pretty(breaks = pretty_log10_breaks()) +
                         scale_y_log10_pretty(breaks = pretty_log10_breaks()) +
                         theme(strip.background = element_blank(),
                               panel.spacing = unit(12, "pt"),
                               strip.text = element_text(margin = margin(0, 0, 2, 0)),
                               plot.margin = margin(7, 12, 7, 7))
                 })

l_ply(gg_list, print)
```

### UMI count distribution

```{r fig.width = 9, fig.height = 3}
message("umis per cell distribution")

gg_list <- dlply(cell_count_summary, .(plot_group),
                 function(df) {
                     gg <- ggplot(df, aes(x = count)) +
                         facet_wrap(~ sample, nrow = 1, scales = "free") +
                         geom_density() +
                         scale_x_log10_pretty(breaks = pretty_log10_breaks()) +
                         coord_cartesian(expand = F) +
                         theme(strip.background = element_blank(),
                               panel.spacing = unit(12, "pt"),
                               strip.text = element_text(margin = margin(0, 0, 2, 0)),
                               axis.ticks.y = element_blank(),
                               axis.text.y = element_blank(),
                               plot.margin = margin(7, 12, 7, 7))
                 })

l_ply(gg_list, print)

```

## UMI threshold

For now just using a threshold of 5 UMIs

```{r}
umi_threshold <- 5

```


## Number of unique features per cell

```{r fig.width=9, fig.height=3}

message("unique features per cell")

features_per_cell <- feature_counts %>%
    filter(sample != "pDNA") %>%
    filter(count >= umi_threshold) %>% 
    group_by(sample, cell) %>%
    summarise(features = n()) %>% 
    ungroup() %>%
    mutate(plot_group = ceiling(as.numeric(factor(sample))/samples_per_row))

total_cells <- features_per_cell %>% 
    group_by(sample, plot_group, features) %>% 
    summarise(n = n()) %>% 
    group_by(sample, plot_group) %>% 
    mutate(pct = percent_format(accuracy = .1)(n/sum(n)))
    

gg_list <- dlply(features_per_cell, .(plot_group),
                 function(df) {
                     gg <- ggplot(df, aes(features)) +
                         geom_histogram(binwidth = 1) +
                         geom_text(aes(x = features, y = n, label = pct),
                                   vjust = -0.2, size = 3,
                                   data = total_cells %>% filter(plot_group == df$plot_group[1])) +
                         facet_wrap(~ sample, nrow = 1, scales = "free") +
                         scale_x_continuous(breaks=1:7, limits = c(0.5,5.5)) +
                         # scale_y_continuous(limits=c(0, 10500)) +
                         theme(strip.background = element_blank(),
                               panel.spacing = unit(12, "pt"),
                               strip.text = element_text(margin = margin(0, 0, 2, 0)),
                               plot.margin = margin(7, 12, 7, 7),
                               axis.text.y = element_text(size = 8))
                 })

l_ply(gg_list, print)


```

## Distribution of cells per feature

```{r fig.width=9, fig.height=3}

message("cells per feature")

pDNA <- feature_counts %>% 
    filter(sample == "pDNA") %>%
    dplyr::select(feature, pDNA = count)
    

cells_per_feature <-
    feature_counts %>%
    filter(sample != "pDNA",
           count >= umi_threshold) %>%
    group_by(sample, feature) %>%
    summarise(cells = n()) %>%
    ungroup() %>%
    complete(crossing(sample, feature=pDNA$feature), fill = list(cells = 0)) %>% 
    full_join(pDNA) %>% 
    mutate(plot_group = ceiling(as.numeric(factor(sample))/samples_per_row))




quantiles <- c(0.1, 0.9)
annot_df <-
    cells_per_feature %>% 
    group_by(sample, plot_group) %>% 
    do(tibble(q = quantiles,
              c = quantile(.$cells, quantiles),
              h_x = min(.$cells),
              h_xend = c,
              h_y = q,
              v_y = 0,
              v_yend = q,
              v_x = c,
              l = paste0(q, " = ", c)))


gg_list <- dlply(cells_per_feature, .(plot_group),
                 function(df) {
                     a <- annot_df %>% filter(plot_group == df$plot_group[1])
                     gg <- df %>% 
                         group_by(sample) %>% 
                         mutate(ecdf = ecdf(cells)(cells)) %>% 
                         # filter(count > 0) %>% 
                         ggplot(aes(x=cells, y=ecdf)) +
                         facet_wrap(~ sample, nrow = 1, scales = "free") +
                         geom_segment(aes(x=h_x,xend=h_xend,y=h_y,yend=h_y),
                                      data=a, lty=2, color="grey70") +
                         geom_segment(aes(x=v_x,xend=v_x,y=v_y,yend=v_yend),
                                      data=a, lty=2, color="grey70") +
                         geom_text(aes(x=h_xend, y=h_y, label=l), vjust=-0.5, hjust=1.25,
                                   data = a, size = 2.5) +
                         geom_line() +
                         scale_y_continuous(limits=c(0,1), breaks=pretty_breaks(5)) +
                         scale_x_continuous(breaks = pretty_breaks(5)) +
                         # annotation_logticks(sides='b') +
                         # scale_x_log10_pretty(limits = c(min(pDNA_counts$count[pDNA_counts$count > 0]),
                         #                                 max(pDNA_counts$count))) +
                         labs(x="Cells per feature", y="Cumulative fraction") +
                             theme(strip.background = element_blank(),
                                   panel.spacing = unit(12, "pt"),
                                   strip.text = element_text(margin = margin(0, 0, 2, 0)),
                                   plot.margin = margin(7, 12, 7, 7))
                 })

l_ply(gg_list, print)


```

## Selection of cells with features relative to pDNA

```{r fig.width=9, fig.height=3}

message("selection")

cells_per_target <- cells_per_feature %>% 
    mutate(target = str_match(feature, snakemake@config$target_regex)[,2]) %>% 
    mutate(nt = ifelse(str_detect(feature, "NonTarget") | str_detect(feature, "sgNT") | str_detect(feature, "non_target"),
                       "non-targeting", "targeting"),
           target = ifelse(is.na(target) | nt == "non-targeting", feature, target),
           cells = ifelse(is.na(cells), 0, cells),
           pDNA = ifelse(is.na(pDNA), 0, pDNA)) %>% 
    group_by(sample, target, nt) %>%
    summarise(pDNA = sum(pDNA),
              cells = sum(cells))
    

n_labels <- c("non-targeting"=1, "targeting"=5)

feature_logFC <- cells_per_target %>%
    group_by(sample) %>%
    mutate(logFC = log2((cells+1)/sum(cells)) - log2((pDNA+1)/sum(pDNA))) %>% 
    group_by(sample, nt) %>% 
    mutate(label = ifelse(min_rank(logFC) <= n_labels[nt] | 
                              min_rank(-logFC) <= n_labels[nt], str_sub(target, 1, 9), "")) %>% 
    ungroup() %>% 
    mutate(plot_group = ceiling(as.numeric(factor(sample))/samples_per_row))
    




gg_list <- dlply(feature_logFC, .(plot_group),
                 function(df) {
                     gg <- ggplot(df, aes(x=pDNA, y=logFC, color=nt, label=label)) +
                         facet_wrap(~ sample, nrow=1, scales="free") +
                         geom_hline(yintercept = 0, color = "grey70", lty=2) +
                         geom_point(size = 1, show.legend = F) +
                         geom_text_repel(size = 2.5, show.legend = F) +
                         scale_x_continuous(labels = format_si()) +
                         labs(x="pDNA",
                              y="log2 Fold Change") +
                         theme(strip.background = element_blank(),
                               panel.spacing = unit(12, "pt"),
                               strip.text = element_text(margin = margin(0, 0, 2, 0)),
                               plot.margin = margin(7, 12, 7, 7))
                 })

l_ply(gg_list, print)

```


```{r fig.width=9, fig.height=4}

if ("Condition" %in% colnames(samplesheet)) {
    conditions_per_row <- 3
    conditions <- samplesheet$Condition %>% set_names(samplesheet$Name)    
    
    combined_cells_per_target <- cells_per_target %>% 
        mutate(condition = conditions[sample]) %>% 
        group_by(condition, target, nt, pDNA) %>%
        summarise(cells = sum(cells)) %>% 
        mutate(plot_group = ceiling(as.numeric(factor(condition))/conditions_per_row))
        
        n_labels <- c("non-targeting"=1, "targeting"=10)
        
        feature_logFC <- combined_cells_per_target %>%
            group_by(condition) %>%
            mutate(logFC = log2((cells+1)/sum(cells)) - log2((pDNA+1)/sum(pDNA))) %>% 
            group_by(condition, nt) %>%
            mutate(label = ifelse(min_rank(logFC) <= n_labels[nt] | 
                                      min_rank(-logFC) <= n_labels[nt], str_sub(target, 1, 9), "")) %>% 
            ungroup() %>% 
            mutate(plot_group = ceiling(as.numeric(factor(condition))/conditions_per_row))
        
        
        gg_list <- dlply(feature_logFC, .(plot_group),
                         function(df) {
                             gg <- ggplot(df, aes(x=pDNA, y=logFC, color=nt, label=label)) +
                                 facet_wrap(~ condition, nrow=1, scales="free") +
                                 geom_hline(yintercept = 0, color = "grey70", lty=2) +
                                 geom_point(size = 1, show.legend = F) +
                                 geom_text_repel(size = 2.5, show.legend = F) +
                                 scale_x_continuous(labels = format_si()) +
                                 labs(x="pDNA",
                                      y="log2 Fold Change") +
                                 theme(strip.background = element_blank(),
                                       panel.spacing = unit(12, "pt"),
                                       strip.text = element_text(margin = margin(0, 0, 2, 0)),
                                       plot.margin = margin(7, 12, 7, 7))
                         })
        
        l_ply(gg_list, print)
        
        
}


```


